File: iOS/Store/Store/Util/DatabaseFunctions.m
#import "DatabaseFunctions.h"
#import "Bridge.h"

@implementation DatabaseFunctions

+ (NSString *) sqlEscapeString: (NSString *) value
{
	return [NSString stringWithFormat: @"\"%@\"", value];
}

+ (NSArray *) sqlEscapeColumns: (NSArray *) columns
{
	NSMutableArray *escapedColumns = [NSMutableArray new];
	for(int i = 0; i < [columns count]; i++)
	{
		NSString *column = [columns objectAtIndex: i];
		NSString *escapedColumn = [self sqlEscapeString: column];
		[escapedColumns insertObject: escapedColumn atIndex: i];
	}
	return escapedColumns;
}

+ (NSString *) createWhereColumn: (NSString *) name
{
	return [NSString stringWithFormat: @"%@ = ?", [self sqlEscapeString: name]];
}

+ (NSDictionary *) parseValues: (NSString *) json
{
	NSDictionary *object = [Bridge parseJSON: json];
	if(object)
	{
		NSMutableDictionary *values = [NSMutableDictionary new];
		for(NSString *column in object)
		{
			NSString *value = object[column];
			NSString *escapedColumn = [DatabaseFunctions sqlEscapeString: column];
			[values setValue: value forKey: escapedColumn];
		}
		return values;
	} else {
		[Bridge log: [NSString stringWithFormat: @"Cannot parse values: %@", json]];
	}
	return nil;
}

+ (NSArray *) parseColumns: (NSString *) json
{
	NSArray *array = [Bridge parseJSON: json];
	if(array)
	{
		NSMutableArray *columns = [NSMutableArray new];
		for(NSString* column in array)
		{
			[columns addObject: column];
		}
		return columns;
	} else {
		[Bridge log: [NSString stringWithFormat: @"Cannot parse *columns: %@", json]];
	}
	return nil;
}

+ (void) parseWhere: (NSString *) json refWhere: (NSString **) where refWhereArgs: (NSArray **) whereArgs
{
	NSDictionary *object = [Bridge parseJSON: json];
	if(object)
	{
		NSMutableString *clause = [NSMutableString new];
		NSMutableArray *args = [NSMutableArray new];
		int i = 0;
		for(NSString *column in object)
		{
			if(i > 0)
			{
				[clause appendString: @" AND "];
			}
			NSString *value = object[column];
			[clause appendString: [self createWhereColumn: column]];
			[args addObject: value];
			i++;
		}
		*where = clause;
		*whereArgs = args;
	} else {
		[Bridge log: [NSString stringWithFormat: @"Cannot parse where: %@", json]];
	}
}

@end

File: Android/store/src/com/dasimple/mobilekit/store/util/DatabaseFunctions.java
package com.dasimple.mobilekit.store.util;

import java.util.Iterator;
import java.util.concurrent.atomic.AtomicReference;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import com.dasimple.mobilekit.bridge.Bridge;

import android.content.ContentValues;

public final class DatabaseFunctions
{
	public static String sqlEscapeString(String value)
	{
		return String.format("\"%s\"", value);
	}
	
	public static String[] sqlEscapeColumns(String... columns)
	{
		String[] escapedColumns = new String[columns.length];
		for(int i = 0; i < escapedColumns.length; i++)
		{
			escapedColumns[i] = sqlEscapeString(columns[i]);
		}
		return escapedColumns;
	}
	
	public static String createWhereColumn(String name)
	{
		return sqlEscapeString(name) + " = ?";
	}
	
	public static ContentValues parseValues(String json)
	{
		try
		{
			JSONObject object = new JSONObject(json);
			ContentValues contentValues = new ContentValues();
			Iterator<String> keys = object.keys();
			while(keys.hasNext())
			{
				String column = keys.next();
				String value = object.optString(column);
				String escapedColumn = sqlEscapeString(column);
				contentValues.put(escapedColumn, value);
			}
			return contentValues;
		}
		catch (JSONException e)
		{
			Bridge.log("Cannot parse values: " + json);
		}
		return null;
	}
	
	public static String[] parseColumns(String json)
	{
		try
		{
			JSONArray array = new JSONArray(json);
			int length = array.length();
			String[] columns = new String[length];
			for(int i = 0; i < length; i++)
			{
				String column = array.optString(i);
				columns[i] = column;
			}
			return columns;
		}
		catch (JSONException e)
		{
			Bridge.log("Cannot parse *columns: " + json);
		}
		return null;
	}
	
	public static void parseWhere(String json, AtomicReference<String> refWhere, AtomicReference<String[]> refWhereArgs)
	{
		try
		{
			JSONObject object = new JSONObject(json);
			int length = object.length();
			String clause = "";
			String[] args = new String[length];
			int i = 0;
			Iterator<String> keys = object.keys();
			while(keys.hasNext())
			{
				if(i > 0)
				{
					clause += " AND ";
				}
				String column = keys.next();
				String value = object.optString(column);
				clause += createWhereColumn(column);
				args[i] = value;
				i++;
			}
			refWhere.set(clause);
			refWhereArgs.set(args);
		}
		catch (JSONException e)
		{
			Bridge.log("Cannot parse where: " + json);
		}
	}
}